packages/backend/src/open_payments/payment/outgoing/service.ts

```diff
@@ -88,22 +91,57 @@ async function getOutgoingPayment(
   }
 }
 
-export interface CreateOutgoingPaymentOptions {
+export interface BaseOptions {
   walletAddressId: string
-  quoteId: string
   client?: string
   grant?: Grant
   metadata?: Record<string, unknown>
   callback?: (f: unknown) => NodeJS.Timeout
   grantLockTimeoutMs?: number
 }
+export interface CreateFromQuote extends BaseOptions {
+  quoteId: string
+}
+export interface CreateFromIncomingPayment extends BaseOptions {
+  incomingPayment: string
+  debitAmount: Amount
+}
+
+export type CreateOutgoingPaymentOptions =
+  | CreateFromQuote
+  | CreateFromIncomingPayment
+
+export function isCreateFromIncomingPayment(
+  options: CreateOutgoingPaymentOptions
+): options is CreateFromIncomingPayment {
+  return 'incomingPayment' in options && 'debitAmount' in options
+}
 
 async function createOutgoingPayment(
   deps: ServiceDependencies,
   options: CreateOutgoingPaymentOptions
 ): Promise<OutgoingPayment | OutgoingPaymentError> {
+  const { walletAddressId } = options
+  let quoteId: string
+
+  if (isCreateFromIncomingPayment(options)) {
+    const { debitAmount, incomingPayment } = options
+    const quoteOrError = await deps.quoteService.create({
+      receiver: incomingPayment,
+      debitAmount,
+      method: 'ilp',
+      walletAddressId
+    })
+
+    if (isQuoteError(quoteOrError)) {
+      throw quoteOrError
```

Comment:
we _could_ also have this method return type also return `QuoteError`, but maybe this is simpler in general

Reply:
hm... I think that's a good suggestion because thats what we're doing elsewhere in the method. Was kinda tricky to figure out _how_ though. Here are the changes [826b48a](https://github.com/interledger/rafiki/pull/2654/commits/826b48a6a006b037f4661fb232996f9938924904) and here is what I was thinking:

Changing the return type to also include `QuoteError` breaks callers several places and requires more complicated checking there. It also generally feels wrong to return a quote error for an outgoing payment. Maybe not awful but I wanted to look for something better.

Additionally, I think the quote error should be preserved (instead of just doing `OutgoingPaymentError.InvalidQuote` for example) since it contains some potentially actionable info (invalid receiver, receive amount, etc.).

In order to do this I added the `QuoteError` enum pairs to the `OutgoingPaymentError` enum and made a `quoteErrorToOutgoingPaymentError` that lets us do:

```ts
    if (isQuoteError(quoteOrError)) {
      return quoteErrorToOutgoingPaymentError[quoteOrError]
    }
```

I think this is probably the best we can do but it also made me think it would be nice if the errors were class-based. That way we could set the details (such as the message) on an instance-by-instance basis instead.